hail/src/main/scala/is/hail/annotations/aggregators/RegionValueApproxCDFAggregator.scala

```diff
@@ -0,0 +1,706 @@
+    this(levels, items, numLevels, helper.dummyValue, helper.dummyValue)
+
+  def copy(): ApproxCDFCombiner[T] =
+    new ApproxCDFCombiner[T](levels.clone(), items.clone(), numLevels, minValue, maxValue)
+
+  def maxNumLevels = levels.length - 1
+  def capacity = items.length
+  def isFull = levels(0) == 0
+
+  def size = levels(maxNumLevels) - levels(0)
+
+  def levelSize(level: Int): Int = levels(level + 1) - levels(level)
+
+  def safeLevelSize(level: Int): Int =
+    if (level >= maxNumLevels) 0 else levels(level + 1) - levels(level)
+
+  def push(t: T) {
+    val bot = levels(0)
+
+    if (bot == capacity) {
+      minValue = t
+      maxValue = t
+    } else if (helper.lt(t, minValue)) {
+      minValue = t
+    } else if (helper.lt(maxValue, t)) {
+      maxValue = t
+    }
+
+    val newBot = bot - 1
+    items(newBot) = t
+    levels(0) = newBot
+
+  }
+
+  def grow(newNumLevels: Int, newCapacity: Int): ApproxCDFCombiner[T] = {
+    require(newNumLevels > maxNumLevels && newCapacity > capacity)
+    val newLevels = Array.ofDim[Int](newNumLevels + 1)
+    val newItems = Array.ofDim[T](newCapacity)
+    val shift = newCapacity - capacity
+    var i = 0
+    while (i <= maxNumLevels) {
+      newLevels(i) = levels(i) + shift
+      i += 1
+    }
+    while (i <= newNumLevels) {
+      newLevels(i) = newCapacity
+      i += 1
+    }
+    System.arraycopy(items, levels(0), newItems, newLevels(0), size)
+
+    new ApproxCDFCombiner[T](newLevels, newItems, numLevels, minValue, maxValue)
+  }
+
+  def clear() {
+    minValue = helper.dummyValue
+    maxValue = helper.dummyValue
+    numLevels = 1
+    var i = 0
+    while (i < levels.length) {
+      levels(i) = items.length
+      i += 1
+    }
+  }
+
+  /* Compact level `level`, merging the compacted results into level `level+1`.
+   * Shift lower levels up to keep items contiguous.
+   */
+  def compactLevel(level: Int, shiftLowerLevels: Boolean = true) {
+    assert(level < maxNumLevels - 1)
+    if (level == numLevels - 1) numLevels += 1
+
+    val bot = levels(0)
+    val a0 = levels(level)
+    val c = levels(level + 1)
+    val d = levels(level + 2)
+
+    val size = c - a0
+    val halfSize = size / 2
+    val adj = size % 2
+
+    val a = a0 + adj
+    val b = a + halfSize
+
+    val sizeBelow = a - bot
+    val sizeAbove = d - c
+
+    if (level == 0) helper.sort(items, a, c)
+    if (sizeAbove == 0) {
+      helper.compactBufferBackwards(items, a, c, items, c)
+    } else {
+      helper.compactBuffer(items, a, c, items, a)
+      helper.merge(items, a, b, items, c, d, items, b)
+    }
+    levels(level + 1) = b
+
+    if (shiftLowerLevels) {
+      // shift up values below
+      if (sizeBelow > 1) {
+        // copy [bot, a) to [bot+halfSize, b)
+        System.arraycopy(items, levels(0), items, bot + halfSize, sizeBelow)
+      } else {
+        // only needs to be done if sizeBelow == 1, but doesn't hurt otherwise
+        items(b - 1) = items(a0)
```

Comment:
Why not just use array copy to copy the one element?

Reply:
Also, what if there's one element in level 0 and an even number of elements in level 1, then `sizeBelow` is `1`, `a0 == a`, and we should actually be copying `a - 1`. Does that never happen?

Reply:
> Why not just use array copy to copy the one element?

If you look at the docstring for `arraycopy`, it does a lot of precondition checking, figuring out if the in and out ranges overlap, etc. The `sizeBelow == 1` case is very common, and I suspect (but haven't checked) that avoiding `arraycopy` is faster in this case.

> Does that never happen?

I don't think that can happen, but I also don't think this code should be relying on that. I think replacing `a0` with `bot` fixes it. Agree?

Reply:
yeah bot seems right